### ISSUE TITLE
Funnel sending `Message`s through `SendMessageWithReplyAsync()` when broadcasting via `NetMQTransport`

### ISSUE THREAD

**Issue by greymistcube** [planetarium/libplanet#1722]
Unlike newly implemented `TcpTransport`, `NetMQTransport`'s broadcasting mechanism implements its own way of sending `Message`s rather than reusing `SendMessageWithReplyAsync()` interface method.

There are several problems with this:

- There is no good reason to have a separate implementation.
- Although the functionality looks the same for sending a `Message` to a `BoundPeer` from the outside, two completely different implementations need to be maintained.
  - As an example, any additional functionality we'd want to attach to the act of sending a `Message`, this needs to be done twice, also in different manner to accommodate each implementation.

I suggest we reroute internally `DoBroadcast()` to `SendMessageWithReplyAsync()` for `NetMQTransport`.

**Comment by greymistcube** [comment]
On further inspection, funneling `DoBroadcast()` to `SendMessageWithReplyAsync()` doesn't seem as simple as it seems. 😶

However, I still believe the act of "sending a message" should ultimately follow a single "controllable" code path, as there is no good reason other than some orchestrated handling of `DealerSocket`s necessary for reusing cached `DealerSocket`s.

At the moment, refactoring our custom `NetMQSocketExtensions` seem to be the easiest way to go about resolving this issue. 😗

**Comment by greymistcube** [comment]
Yet, on another further inspection, current usage of `DealerSocket` seems somewhat improper. I didn't know we were using a [customized NetMQ](PlanetariumNetMQ) when sending messages to "force" asynchronous sending. After some research on [NetMQ] and [ZeroMQ], the custom implementation seems to be at odds with how these library are "supposed" to be used.

[PlanetariumNetMQ]: https://github.com/planetarium/netmq
[NetMQ]: https://github.com/zeromq/netmq
[ZeroMQ]: https://zeromq.org

**Comment by greymistcube** [comment]
Eh, thought this was going to be a quick fix but is getting out of hand. I suppose some explanation is in order. Hopefully this doesn't turn into a long rant. 😶

To clarify, the main goal is to have a reliable `ITransport` implementation, as the general consensus still seems to be that neither `TcpTransport` nor `NetMQTransport`, two different implementations of `ITransport`, can be mostly trusted. As one of the lowest level component for [libplanet] library, unreliability in transport layer results in overall high cost for troubleshooting and patching [libplanet]. As only `NetMQTransport` is deployed at the moment, the transport layer henceforth will refer to `NetMQTransport`.

The first step is to determine roughly how reliable, or rather unreliable, the transport layer is. With the codebase as it is, the fastest solution would be to add new logs for incoming and outgoing messages everywhere. It turns out there are two separate code paths for each incoming messages and outgoing messages. Seems like a no big deal to add four more logs, but I was afraid with a fast rollout of a highly visible metric, it would result in subsequent duct-taping of solutions on top of already unnecessarily complex, i.e. business as usual.

The strategy was to do a minimal refactoring, or at least what it seemed like on the surface level at the time, to somewhat mitigate my concern. As mentioned in previous comments, issues ran deeper than I originally thought.

**Rant start**

I saw the underlying problem to be several-fold:

- Unnecessarily complex implementation of the transport layer.
- Lack of concern separation between the transport layer and other components.
- Obfuscation of possible problems in [NetMQ].

These are generally hinderances that normally arise in long development by multiple contributors. However, if a simple subtask of funneling a couple message communication channels into one turns out to be this complicated, I can hardly imagine the transport layer becoming sufficiently reliable without giving it a closer attention. So I started to chip away what I can here and there, all the while continually discovering more possible issues. What's more, even with gradual changes, due to its nature, any amount of behavioral change in the transport layer needs to be tested on `9c-main`, which added more developmental cost than originally anticipated.

**Rant end**

Pertaining to the issue at hand, from what I have gathered so far, at the time of this writing, usage of `TrySendMultipartMessageAsync()` is in direct conflict with [NetMQ]'s intended usage. This seems to have had an unintended substantial amount of ripple effect on the transport layer resulting in fairly heavy unintuitive workarounds. I believe strongly coupled with this unorthodox implementation is the issue of the management of reusing `DealerSocket`s. Although further investigation is needed, simply using `TrySendMultipartMessage()` seem to work fairly well until the node runs into socket issues, which I suspect is due to the management of `DealerSocket`s requiring a different management strategy. My previous failed attempts, including #1711, came from not sufficiently understanding [NetMQ] and the workarounds gradually built up over time for the transport layer. I feel fairly confident at this point that switch to the vanilla flavored [NetMQ] is possible.

The general direction I have in mind for this issue is the following:

- Replace `TrySendMultipartMessageAsync()` with `TrySendMultipartMessage()`.
- Implement a new `Exception` such as `MessageSendFailedException` for finer flow control.
- Change `DealerSocket`s management accordingly.

This is based on the premise that [NetMQ] works fairly well. In case of [NetMQ] failure, this should be handled more explicitly, not as some unknown possible `Exception` so that the more abstract level logic can continue on more gracefully.

[libplanet]: https://github.com/planetarium/libplanet
[NetMQ]: https://github.com/zeromq/netmq
